tests/storage/snapshot.go

```diff
@@ -436,14 +437,16 @@ var _ = SIGDescribe("VirtualMachineSnapshot Tests", func() {
 
 				//update vm to make sure vm revision is saved in the snapshot
 				By("Updating the VM template spec")
-				newVM, err := virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &metav1.GetOptions{})
-				Expect(err).ToNot(HaveOccurred())
+				err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
What about using a patch?

Reply:
I'm not an expert, but I think patches are suppose to be used in fields which there is no potential for collisions. E.g. labels, annotations or just parts that are under one component/routine responsibility and no one else writes there.

I'm not sure if we can assume this, or want to assume it.
I personally have no strong opinion here, just a bit worried patching is not overused , as it can introduce worse problems.

Reply:
IIUC if the collision occurs on the `Requests` fields then update would be better, otherwise patch can also work. However from the error message its not clear which specific fields were outdated

Reply:
Not sure I follow but here I see we only update the memory(in spec). I don't know what would be a better candidate for a patch.

Reply:
Lets try to put the known parts here and then you tell me your thoughts:
- Update vs Path: Update has a collision detection, which makes sure an update is not lost. Patch does not have it, at least not that I know of.
- The test in question assumes the change done has indeed succeeded (in other words not lost), therefore the options are either to keep using update, use patch and then re-fetch the VMI to see it is with the new value or be optimistic and hope the patch succeeded, because no one else should have touched this value.
I am usually not optimistic :)
I can argue that kubevirt may update the VMI status using Update, and therefore will override any patching done in parallel. We can delve into this deeper if you want.

Reply:
@EdDev I've found [this](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#common-architecture-flaws-to-avoid) doc in our project

Reply:
> Now we are left only to decode the reasoning behind this recommendation.
> Do you understand what is the reasoning?

Here's some reasoning.
* An update can be lossy if the client updating the resource is not aware of new fields in the API. that's why only the controller that "owns" the object should be responsible for issuing an Update.
* The other reason is collisions. Two controllers attempting to perform updates can collide and cause lots of retries, which is inefficient. 

If the controller loop happens to be in the same process as the other contorller loop that owns the object, then we're primarily avoiding collisions by using a Patch, which is important.

In combination with a `test` operation, Patch is completely safe and preferred when outside of the owning controller loop. The `test` operation should be used before the `add`, `replace`, `delete`, operation in order to verify that the fields the patch is mutating have not changed on the backend from the cached version of the resource being used.

This is why you'll see a `test` op coupled with most of our patch operations.

Reply:
now all that said... This test code works without race now the way you have it.

A patch would be preferred, but retrying the Update also technically works here (but less efficiently). efficiency isn't a huge deal for e2e tests. I'd feel differently if this was a controller loop though.

Reply:
> Here's some reasoning.
> 
>     * An update can be lossy if the client updating the resource is not aware of new fields in the API. that's why only the controller that "owns" the object should be responsible for issuing an Update.

This is a strong one, thanks!
Only the owner is in sync with the current API, others may use a different API.

I guess we do [have a problem with the virt-handler](https://github.com/kubevirt/kubevirt/blob/ceefa1a087a1ae09eb1e9ba04e9c434aae361617/pkg/virt-handler/vm.go#L1197) then, I wander how come we never experienced problems there.

> 
>     * The other reason is collisions. Two controllers attempting to perform updates can collide and cause lots of retries, which is inefficient.
> 
> 
> If the controller loop happens to be in the same process as the other contorller loop that owns the object, then we're primarily avoiding collisions by using a Patch, which is important.
> 
> In combination with a `test` operation, Patch is completely safe and preferred when outside of the owning controller loop. The `test` operation should be used before the `add`, `replace`, `delete`, operation in order to verify that the fields the patch is mutating have not changed on the backend from the cached version of the resource being used.
> 
> This is why you'll see a `test` op coupled with most of our patch operations.

Ok, this sounds very nice and makes total sense.
I guess the `test` operation fits well when we know exactly which area changed (once we reach the location when we need to perform the change).
Will it also fit when we want to update a large subset of the manifest? (e.g. `vmi.Status`)

> now all that said... This test code works without race now the way you have it.
> 
> A patch would be preferred, but retrying the Update also technically works here (but less efficiently). efficiency isn't a huge deal for e2e tests. I'd feel differently if this was a controller loop though.

I find tests to be an excellent entry point for learning a new project.
Therefore, my state of mind is to express the right way of doing things in tests, considering it documentation-by-example.

As such, I am going to take this important feedback and try to move all `Update`/s that change specific fields in tests to `Patch`. Starting with this one.

If this test is a pain (continuing to heavily flake), I leave it to @xpivarc / @enp0s3 to unhold and let in get in. I can just change it after.

Thank you @davidvossel for this great info!

Reply:
/unhold
I will take the stability first then. Lets optimize with test & patch in a follow up.

Reply:
@davidvossel 
Thank you for the comprehensive explanation.

Note: I just wanted to know if this was considered and if so why it wasn't used. IMHO we could improve the test by using patch in a way that I often see we use update incorrectly and IMO patch seems to be easier to use right.